### ISSUE TITLE
Settings tab improvements

### ISSUE THREAD

**Issue by PilzAdam** [minetest/minetest#3266]
The todo list from the pull request:
- [x] `flags` type for e.g. mapgen flags with a list of possible values.
- [x] Fancy input for `flags` type; a list of checkboxes with all available options would be cool here.
- [x] `noise_params` type; since they are always the same structure and they are excessively used I think they deserve a special treatment in the GUI.
- [x] Fancy input for `noise_params` type.
- [x] Input validation for `float` and `int`; check if entered string is really a number and check min and max.
- [ ] The engine supports arbitrary structs in the settings, but there is no setting that uses this (except noise params, but they get a special treatment); this is low priority.
- [ ] Maybe a fancier GUI.

Additional improvements:
- [x] Parse additional `settingtypes.txt` for all installed mods and games and listing them in "Mods" and "Games" categories with their respective names as sub-categories.
- [x] Fix editing of secure.\* settings.
- [ ] Add command line options to automagically generate `minetest.conf.example` (and translation files) for the engine and mods / games.
- [x] Add a search field for settings.
- [ ] Expand / collapse all

**Comment by kilbith** [comment]
Another important thing for not wasting time : the subtrees should stay unrolled after editing a setting (but that's directly related with formspec's behavior).

**Comment by PilzAdam** [comment]
@kilbith that is a generic formspec feature and thus out-of-scope for this. @kahrl is currently working on it, though.

**Comment by kahrl** [comment]
@kilbith Fixed by c4d18623440b58b9f90629c798687608bc32c6da.

Another issue I just found: Trying to modify a secure setting (e.g. `secure.enable_security`, `secure.trusted_mods`) causes an error message like:
```
Runtime error from mod '' in callback handleMainMenuButtons(): Attempt to set secure setting.
```

**Comment by PilzAdam** [comment]
@kahrl builtin should be generally trusted and it should be allowed to modify these settings. The security code should be modified to allow that.

**Comment by kilbith** [comment]
The users are generally unfavorable to the complexity of the new tree system and some legit reasons are detailled here : https://forum.minetest.net/viewtopic.php?f=18&t=13427

Consequently, the user-friendly interface should be preserved by putting back the old setting tab, with the addition of a [Advanced settings] button linking out to the new configuration.

**Comment by paramat** [comment]
> Fancy input for float and int; if max and min are specified then use a slider.

No thanks, i usually need to input a precise value.

**Comment by kilbith** [comment]
https://github.com/blockplanet/blockplanet/commit/56d545b2a1f55bca827673e6537cd28499282a50

**Comment by red-001** [comment]
The new settings tab is very confusing and new player will find it hard to use. Maybe the old settings tab could be returned and this used as an advance settings tab or something of that sort.

**Comment by RHRhino** [comment]
@red-001 the new settings are work in progress and nothing final.

**Comment by PilzAdam** [comment]
I added "Add a search field for settings." to the first post.

**Comment by paramat** [comment]
@PilzAdam there are PRs for the return of the old settings in parallel, i think the branching structure needs rework to make often-used settings (the settings in the previous settings menu) faster to access. Perhaps the first branching should not be 'server, client etc' as that's not obvious to many users.

Just to document another issue: https://github.com/minetest/minetest/pull/3402
'Allow a blank default for mapgen flags' (fix is coded but not merged) and 'Allow the input of a blank field for mapgen flags'.

**Comment by Rogier-5** [comment]
I don't exactly know where the problem originates, but when editing a setting, after entering an invalid value (number out of range, bad flag) and pressing 'save', the 'save' button cannot be pressed again after the input has been corrected. The only way to accept the change then is to press enter.

Also, when a value is accepted using the enter key (whether it was bad before or not), the change-setting dialog closes and reopens immediately.

**Comment by 0-afflatus** [comment]
The top level "Client and Server" is next to meaningless. Ordinary players don't use this terminology. I would unpack the second level as "controls, "network, "graphics" etc are the headings I would expect. Rename "Client and Server" to "General" or something similar. Better to have a longer list of headings containing fewer options in the top level.

**Comment by paramat** [comment]
@PilzAdam is there any issue here that needs attention before release in a month? Any possible blockers?

**Comment by PilzAdam** [comment]
@paramat
> Fix editing of secure.* settings.

This could be a blocker.

> Add a search field for settings.

This would greatly improve usability, but I think this isn't a blocker.

**Comment by paramat** [comment]
Also #3808

**Comment by paramat** [comment]
> Fix editing of secure.* settings.
> This could be a blocker.

@PilzAdam are you able to work on this? feature freeze is tomorrow.

**Comment by PilzAdam** [comment]
@paramat I have no idea how the `secure.*` settings work, so someone else should do this.

**Comment by paramat** [comment]
@PilzAdam any idea who?

**Comment by paramat** [comment]
Thanks, didn't see the checkbox. Removing milestone.

**Comment by paramat** [comment]
Mgv5 ground noise cannot be set, and is therefore missing from minetest.conf when it is auto-generated, despite it being settable in .conf using a table form:
```
mgv5_np_ground = {
    offset      = 0
    scale       = 40
    spread      = (80, 80, 80)
    seed        = 983240
    octaves     = 4
    persistence = 0.55
    lacunarity  = 2.0
    flags       = "eased"
}
```
@PilzAdam please can you work on this? Or should someone else do so?

**Comment by DS-Minetest** [comment]
Shouldn't be the most of the list be done by it's own and not all together?
Eg. it would be extremely useful to have
> Add a search field for settings.

with eg.172 objects in mod folder but this has not much to do with the other things, I think.

**Comment by rubenwardy** [comment]
> Add a search field for settings

Done

**Comment by paramat** [comment]
We need to re-organise the setting tree to be more usable, many players will not know what is 'server' or 'server/singleplayer' etc. I still don't know where mapgen will be.
I suggest we move many sections to the first level to be immediately visible, especially for things players want to acess, like mapgen, graphics and shaders. I will move 'mapgen' to the first level.

**Comment by paramat** [comment]
See #6247 (merged) and #6326

**Comment by paramat** [comment]
#6399 merged, one more box ticked.

**Comment by CoderForTheBetter** [comment]
I'm thinking of doing some of the work on this list. Anyone know what this is specifically trying to say, maybe @paramat?: 

"The engine supports arbitrary structs in the settings, but there is no setting that uses this (except noise params, but they get a special treatment); this is low priority."

Should we get rid of them? Support them better? What?

Thanks for any clarification.

**Comment by paramat** [comment]
I think that is suggesting that advanced settings could be made to support arbitrary structs, but this is low priority because no settings use arbitrary structs.
So there is nothing to remove and not much point adding this support yet, don't worry about it for now.

**Comment by paramat** [comment]
More boxes have been ticked, what is left is low priority.

**Comment by Montandalar** [comment]
I'm not sure the human-readable setting names for mods are translatable currently, which would also be an improvement. Either that or I don't have any mods installed that do that.
